File: pokeredus/pokeredus/classes/natures.py
"""
NatureClass — a Pokémon nature with +10% / -10% stat modifiers.
"""

from __future__ import annotations
from dataclasses import dataclass


@dataclass
class NatureClass:
    """A single nature (e.g. Jolly, Adamant, Timid)."""

    name: str
    increased_stat: str | None = None   # "atk", "def", "spa", "spd", "spe"
    decreased_stat: str | None = None   # same, or None for neutral natures

    @property
    def id(self) -> str:
        return self.name.lower()

    @property
    def is_neutral(self) -> bool:
        return self.increased_stat is None and self.decreased_stat is None

    def modifier(self, stat: str) -> float:
        """Return 1.1, 0.9, or 1.0 for the given stat key."""
        if stat == self.increased_stat:
            return 1.1
        if stat == self.decreased_stat:
            return 0.9
        return 1.0

    # ── serialization ───────────────────────────────────────────────
    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "increased_stat": self.increased_stat,
            "decreased_stat": self.decreased_stat,
        }

    @classmethod
    def from_dict(cls, data: dict) -> NatureClass:
        return cls(
            name=data["name"],
            increased_stat=data.get("increased_stat"),
            decreased_stat=data.get("decreased_stat"),
        )

    def __repr__(self) -> str:
        if self.is_neutral:
            return f"NatureClass({self.name!r}, neutral)"
        return f"NatureClass({self.name!r}, +{self.increased_stat} -{self.decreased_stat})"


# ── All 25 natures ──────────────────────────────────────────────────
STANDARD_NATURES: list[NatureClass] = [
    NatureClass("Hardy"),
    NatureClass("Lonely",   "atk", "def"),
    NatureClass("Adamant",  "atk", "spa"),
    NatureClass("Naughty",  "atk", "spd"),
    NatureClass("Brave",    "atk", "spe"),
    NatureClass("Bold",     "def", "atk"),
    NatureClass("Docile"),
    NatureClass("Impish",   "def", "spa"),
    NatureClass("Lax",      "def", "spd"),
    NatureClass("Relaxed",  "def", "spe"),
    NatureClass("Modest",   "spa", "atk"),
    NatureClass("Mild",     "spa", "def"),
    NatureClass("Bashful"),
    NatureClass("Rash",     "spa", "spd"),
    NatureClass("Quiet",    "spa", "spe"),
    NatureClass("Calm",     "spd", "atk"),
    NatureClass("Gentle",   "spd", "def"),
    NatureClass("Careful",  "spd", "spa"),
    NatureClass("Quirky"),
    NatureClass("Sassy",    "spd", "spe"),
    NatureClass("Timid",    "spe", "atk"),
    NatureClass("Hasty",    "spe", "def"),
    NatureClass("Jolly",    "spe", "spa"),
    NatureClass("Naive",    "spe", "spd"),
    NatureClass("Serious"),
]

File: packages/core/src/classes/natures.ts
export class NatureClass {
  constructor(
    readonly name: string,
    readonly increased_stat: string | null = null,
    readonly decreased_stat: string | null = null,
  ) {}

  get id(): string {
    return this.name.toLowerCase();
  }

  get is_neutral(): boolean {
    return this.increased_stat === null && this.decreased_stat === null;
  }

  modifier(stat: string): number {
    if (stat === this.increased_stat) return 1.1;
    if (stat === this.decreased_stat) return 0.9;
    return 1.0;
  }

  toDict(): Record<string, unknown> {
    return {
      name: this.name,
      increased_stat: this.increased_stat,
      decreased_stat: this.decreased_stat,
    };
  }

  static fromDict(data: { name: string; increased_stat?: string | null; decreased_stat?: string | null }): NatureClass {
    return new NatureClass(data.name, data.increased_stat ?? null, data.decreased_stat ?? null);
  }
}

export const STANDARD_NATURES: NatureClass[] = [
  new NatureClass('Hardy'),
  new NatureClass('Lonely', 'atk', 'def'),
  new NatureClass('Adamant', 'atk', 'spa'),
  new NatureClass('Naughty', 'atk', 'spd'),
  new NatureClass('Brave', 'atk', 'spe'),
  new NatureClass('Bold', 'def', 'atk'),
  new NatureClass('Docile'),
  new NatureClass('Impish', 'def', 'spa'),
  new NatureClass('Lax', 'def', 'spd'),
  new NatureClass('Relaxed', 'def', 'spe'),
  new NatureClass('Modest', 'spa', 'atk'),
  new NatureClass('Mild', 'spa', 'def'),
  new NatureClass('Bashful'),
  new NatureClass('Rash', 'spa', 'spd'),
  new NatureClass('Quiet', 'spa', 'spe'),
  new NatureClass('Calm', 'spd', 'atk'),
  new NatureClass('Gentle', 'spd', 'def'),
  new NatureClass('Careful', 'spd', 'spa'),
  new NatureClass('Quirky'),
  new NatureClass('Sassy', 'spd', 'spe'),
  new NatureClass('Timid', 'spe', 'atk'),
  new NatureClass('Hasty', 'spe', 'def'),
  new NatureClass('Jolly', 'spe', 'spa'),
  new NatureClass('Naive', 'spe', 'spd'),
  new NatureClass('Serious'),
];
